pkg/services/dashboards/database/database.go

```diff
@@ -1055,3 +1055,90 @@ func readQuotaConfig(cfg *setting.Cfg) (*quota.Map, error) {
 	limits.Set(orgQuotaTag, cfg.Quota.Org.Dashboard)
 	return limits, nil
 }
+
+func (d *dashboardStore) SaveDashboardWithMetadata(ctx context.Context, msg string, dash *dashboards.Dashboard, provisioning *dashboards.DashboardProvisioning) (*dashboards.Dashboard, error) {
+	var result *dashboards.Dashboard
+	var err error
+
+	err = d.store.WithTransactionalDbSession(ctx, func(sess *db.Session) error {
+		result, err = saveDashboardWithUID(sess, msg, dash, d.emitEntityEvent())
+		if err == nil && provisioning != nil {
+			provisioning.DashboardID = result.ID
+			return saveProvisionedData(sess, provisioning, result)
+		}
+		return err
+	})
+	return result, err
+}
+
+func saveDashboardWithUID(sess *db.Session, msg string, dash *dashboards.Dashboard, emitEntityEvent bool) (*dashboards.Dashboard, error) {
+	var existing dashboards.Dashboard
+	dashWithUIDExists, err := sess.Where("uid=? AND org_id=?", dash.UID, dash.OrgID).Get(&existing)
+	if err != nil {
+		return nil, err
+	}
+
+	var affectedRows int64
+	if dashWithUIDExists {
+		dash.ID = existing.ID
+		dash.SetVersion(existing.Version + 1)
+		affectedRows, err = sess.MustCols("folder_id").ID(dash.ID).Update(dash)
+	} else {
+		dash.SetVersion(1)
+		metrics.MApiDashboardInsert.Inc()
+		affectedRows, err = sess.Insert(dash)
+	}
+
+	if err != nil {
+		return nil, err // error inserting rows
+	}
+
+	if affectedRows == 0 {
+		return nil, dashboards.ErrDashboardNotFound
+	}
+
+	dashVersion := &dashver.DashboardVersion{
+		DashboardID:   dash.ID,
+		ParentVersion: existing.Version,
+		//	RestoredFrom:  cmd.RestoredFrom, ???
+		Version:   dash.Version,
+		Created:   time.Now(),
+		CreatedBy: dash.UpdatedBy,
+		Message:   msg,
+		Data:      dash.Data,
+	}
+
+	//---------------------------------------------
+	// Past here is the same as saveDashboard(...)
```

Comment:
nit: Could possibly refactor the common code if its exactly the same between the 2 code paths.

Reply:
That makes sense, but will do it after the POC (hoping to avoid any drift as the services seem to keep getting refactored)